requests-for-comments/2024-01-23-project-environments.md

```diff
@@ -0,0 +1,192 @@
+
+To support the most common workflow of drafting dashboards/etc. in dev/staging and then "promoting" to production, entities such as dashboards would be environment-specific by default, but could be rolled out to the whole project with two clicks. When rolled out, the entity's dependencies would be made project-wide as well (e.g. for a cohort, all actions and other cohorts used in the definition would be instantly made project-wide as well). We'd introduce checks to make sure you can't make an entity environment-specific if it's currently being used project-wide. Sharing in this way would involve no copying though, so there would be no effort needed to sync things.
+
+To support large customers with access control needs, we'd need to add environment-specific project membership to the existing project access control system.
+
+To support the common feature flags workflow of rollout conditions being different between dev and production, we'd make flag rollout groups always environment-specific.
+
+To ensure straightforward yet complete separation of data at the ingestion level (incl. persons), *as well as* steady querying performance, each environment would have its own `team_id`. Due to this requirement, environments would have to be predefined by users and not created at ingestion time, thus each environment would also have its own SDK key (aka "project API key" currently). That seems like a safe choice in terms of the UX.
+
+By default, a project would only have one environment implicitly named "Production". A "New environment" button would allow adding a new one, with the environments feature explained in the modal. The current environment would then be selectable from the project selector, roughly looking like this (courtesy of Cory):
+
+Extra internal requirement: Because Team Pipeline plans to implement multiple SDK keys per project, it would be simplest to move SDK keys to their own model, rather than moving to the environment model.
+
+<img width="939" alt="image" src="https://github.com/PostHog/meta/assets/4550621/ec966c6b-0a1c-4cb0-a593-0234f4807ab1">
+
+### Downsides
+
+Of course, nothing comes for free: adding evironments has its own downsides. Here, it's the amount of effort required – this feature would touch every part of the app, plus ingestion. It would realistically be a few person-sprints of work (as always, it's hard to estimate accurately).
+
+### Pricing
+
+Environments should start out as a paid feature, since they're key for professional users, and more of a nice-to-have for hobbyists. I would suggest it falling under the same restrictions as multiple projects, i.e. multiple environments being available after putting in a card.
+
+### Rough scope of work
```

Comment:
While I appreciate the zero-downtime transition you're describing, I'm really concerned about the bugs that will arise from having the `team_id` column's meaning being inconsistent between tables (it says the project on event props, but now means environment in apps). The fact that single-environment projects would not be impacted (because env ID == project ID) will increase the risk of such bugs leaking through QA and only manifesting on multi-env teams (our big customers).

I know that renaming columns in PG and CH is a locking operation and a non-backwards-compat change, but I feel like a one-hour maintenance window (where we still capture to kafka and serve decide from redis) to rename all `team_id` columns into `env_id` an `project_id` would save us a lot of headache and bugs down the line. [pgroll](https://github.com/xataio/pgroll) could enable a zero-downtime rename, but I don't think it would play well with future django migrations.

Reply:
Yeah, there's a bunch of cleanup that we should do to avoid confusion in the future 😅 Starting with `posthog_team`, which should be `posthog_project`. It's an exceptionally tricky problem though, so I propose we do it as cleanup at the end of the project.
pgroll sounds great actually, why do you say it wouldn't play well with Django migrations?

Reply:
heavy +1 for this. Let's start with any new columns we add shouldn't use 'team' in it's name (currently 
> `posthog_environment` table, columns int `id`, `team_id`,

Reply:
> It's an exceptionally tricky problem though, so I propose we do it as cleanup at the end of the project.

It is my experience that such cleanup will not be prioritized if not in the critical path, so we should either acknowledge the tech debt we're creating and discuss how to mitigate its impact, or commit to not creating it in the first place.

> pgroll sounds great actually, why do you say it wouldn't play well with Django migrations?

We'd need to create a django migration for the renames, that is applied on new instances but neutralized on the cloud regions, and apply it through pgroll on prod. I see many ways for us to shoot ourselves in the foot with this and create an outage.

Ellie campaigned for us to move away from django migrations and apply schema migrations in a out-of-band zero-downtime process and I support this, especially as we're decoupling new domains (new webhook delivery, next person model) from the django ORM and having a common migration tooling would be great. I'm afraid that would not be a short-term project though.

Reply:
Hmm, I think the critical cleanup part here is renaming ClickHouse `team_id` to `environment_id`. This can result in really bad confusion without cleanup.
The rest is a nice to have because we'll still use "team" for "project".

Reply:
As for naming columns in new models `project_id` instead `team_id`, I'm worried this would be _even_ more confusing until we do the complete cleanup on an undefined timeline. When we do the complete cleanup, it will be marginal effort to rename one more column, so I'll go for `team_id` for continued consistency when meaning projects

Reply:
>  The rest is a nice to have because we'll still use "team" for "project".

We agreed that apps, batch export and data warehouse entities stay at the environment level, so at least in these tables `team_id` will mean environment unless we migrate them. Right?

Reply:
Instead of `team_id` == project and create a new environment entity, I'd love to evaluate what schema changes would be needed if we say `team_id` == environment, and we create a new project entity on top of it. This means that all existing entities stay linked to the environment and can be promoted to shared across the whole project?

Reply:
Just to make this more clear here's what the current proposal is:
<img width="1034" alt="Screenshot 2024-03-01 at 16 11 49" src="https://github.com/PostHog/meta/assets/890921/0664a5e2-d701-42cd-ab42-f0f82599eb5c">

here's an alternative we could consider instead

<img width="1063" alt="Screenshot 2024-03-01 at 16 11 41" src="https://github.com/PostHog/meta/assets/890921/f599fc37-acc6-4fa4-a5fb-5702cfa4ffa1">

Reply:
Ah, I agree that this is way better. So actually in this scenario the "sources" table is orthogonal, because all of ingestion will still use `posthog_team.api_token` (which will mean project API keys simply become environment API keys for the time being, with no extra changes). Will rewrite the plan to do things this way

Reply:
Indeed, we keep a single token per environment for now, then we'll tackle the per-source-token later this year.

The one thing that's not really clear yet is the eventdefinition/propertydefinition/eventproperty table group: they store a mix of:
  - user-defined stuff (descriptions, verified) that makes sense at the project level
  - all the names that plugin-server sees, that might be an issue: your staging data might mess up the dropdown for prod events

Reply:
updated plan
<img width="1010" alt="Screenshot 2024-03-01 at 17 05 06" src="https://github.com/PostHog/meta/assets/890921/6d19b99c-6593-41ec-8531-ad33f8606035">

Reply:
Implemented this plan as new proposed solution D. There are three differences from the graph above in the current scope of work I wrote out:
1. Insights/dashboards/actions/etc. only have `is_project_wide`, since `project_id` can be derived with a quick B-tree join on `posthog_team`.
2. We're going to leave all the settings on the environment level initially.
3. We're not going to `AggregatingMergeTree` as part of _this_ project (but the door's open if you want to try that later).

Reply:
What do you mean by item 3? Is it the revamp of storage for event & property names? Agreed that this is out of scope for this RFC.

Reply:
Yeah yeah, it's the blue post-it on the chart, just clarifying this is out of scope